### ISSUE TITLE
[4.x]: Add ability to sync address book changes to carts using them as source addresses 

### ISSUE THREAD

**Issue by nfourtythree** [craftcms/commerce#3178]
### Description
If a customer has created a cart by using an address or addresses from their address book it is reasonable to think that changes made to these addresses would also be applicable to the cart.

We want to ensure we do not change any completed orders. We need to keep the separation that currently exists in terms of the address model attached to the cart separate from the address owned by the order. This is to make sure that features regarding order editing and historic order data (post-order complete) continue to work as excepted.

Due to the nature of this, we may need to put this under a setting for the time being to avoid any unexpected behavioural changes. Something along the lines of `syncAddressBookChanges`. The setting being `true` would likely become the default in the next major release.

**Comment by bossanova808** [comment]
I think there are two key functional issues to address.  These of course only apply to users with accounts who are logged in (for us this is about 80% of our orders).

1. Logged in users, using saved addresses, expect changes to those to be propagated back to their address book.  This is the behaviour of just about every shopping cart on the planet, and the natural expectation of the user when editing an existing thing (in just about any computing context).  If you do want a new address created from a save one, then a site could offer a 'save as' type option.

2. Logged in users (and new users who choose to register an account at checkout) - expect new addresses they use to be saved for re-use (that, and order history, are really the two reasons to ever both with having accounts on e-commerce sites).  Again, this should be the default - with an opt-out for those that don't want addresses saved (or, in our case, we offer a delete button on the addresses when they are shown later, to make this simple and reduce the checkout question fatigue).

These really should be the default behaviour as it matches the established patterns of online shopping, and indeed the historically established patterns of Commerce itself.  

There _may_ be rare scenarios where those behaviours are not desired (although honestly, I am struggling to think of any such case - e.g. two people ordering with the same address but different name -> they would just create a new address with the other name, right?).  I guess an opt out system might make sense, if you care to implement and support that.  But neither of these things should require custom module code, as they currently now do.

**Comment by nfourtythree** [comment]
Commerce's solution for this have now been merged in ready for the next release. 

More discussion on this topic is on the pull request: https://github.com/craftcms/commerce/pull/3182

Along with all of this we have added and issue to the docs to look into improving the documentation helping developers create a coherent and robust checkout when building with Commerce (https://github.com/craftcms/docs/issues/509).

Thanks!
